src/main/scala/decider/TermToZ3APIConverter.scala

```diff
@@ -0,0 +1,525 @@
+
+      case PredicateLookup(id, psf, args) =>
+        val snap: Term = toSnapTree(args)
+        createApp("$PSF.lookup_" + id, Seq(psf, snap), term.sort)
+
+      case PredicateTrigger(id, psf, args) =>
+        val snap: Term = toSnapTree(args)
+        createApp("$PSF.loc_" + id, Seq(PredicateLookup(id, psf, args), snap), term.sort)
+
+      case PredicatePermLookup(predname, pm, args) =>
+        val snap: Term = toSnapTree(args)
+        createApp("$PSF.perm_" + predname, Seq(pm, snap), term.sort)
+
+      /* Other terms */
+
+      case First(t) => ctx.mkApp(firstFunc, convertTerm(t))
+      case Second(t) => ctx.mkApp(secondFunc, convertTerm(t))
+
+      case bop: Combine =>
+        ctx.mkApp(combineConstructor, convertTerm(bop.p0), convertTerm(bop.p1))
+
+      case SortWrapper(t, to) =>
+        createApp(convertId(SortWrapperId(t.sort, to)), Seq(t), to)
+
+      case Distinct(symbols) =>
+        ctx.mkDistinct(symbols.map(s => ctx.mkConst(convertId(s.id), convertSort(s.resultSort))).toSeq: _*)
+
+      case Let(bindings, body) =>
+        convert(body.replace(bindings))
+
+      case _: MagicWandChunkTerm
+         | _: Quantification =>
+        sys.error(s"Unexpected term $term cannot be translated to SMTLib code")
+    }
+    res
+  }
+
+  @inline
+  protected def createApp(functionName: String, args: Seq[Term], outSort: Sort): Z3Expr = {
+    ctx.mkApp(getFuncDecl(functionName, outSort, args.map(_.sort)), args.map(convertTerm(_)): _*)
+  }
+
+  def getFuncDecl(name: String, resSort: Sort, argSorts: Seq[Sort]): Z3FuncDecl = {
+    val existingEntry = funcDeclCache.get((name, argSorts, resSort))
+    if (existingEntry.isDefined)
+      return existingEntry.get
+    val res = ctx.mkFuncDecl(name, argSorts.map(a => convertSort(a)).toArray, convertSort(resSort))
+    funcDeclCache.update((name, argSorts, resSort), res)
+    res
+  }
+
+
+  @inline
+  protected def createSMTApp(functionName: String, args: Seq[Term], outSort: Sort): Z3Expr = {
+    // workaround: since we cannot create a function application with just the name, we let Z3 parse
+    // a string that uses the function, take the AST, and get the func decl from there, so that we can
+    // programmatically create a func app.
+    val decls = args.zipWithIndex.map{case (a, i) => s"(declare-const workaround${i} ${smtlibConverter.convert(a.sort)})"}.mkString(" ")
+    val funcAppString = s"(${functionName} ${(0 until args.length).map(i => "workaround" + i).mkString(" ")})"
+    val assertion = decls + s" (assert (= ${funcAppString} ${funcAppString}))"
+    val workaround = ctx.parseSMTLIB2String(assertion, null, null, null, null)
+    val app = workaround(0).getArgs()(0)
+    val decl = app.getFuncDecl
+    val actualArgs =  if (decl.getArity > args.length){
+      // the function name we got wasn't just a function name but also contained a first argument.
+      // this happens with float operations where functionName contains a rounding mode.
```

Comment:
Can this happen for other reasons (e.g. accidentally calling a function with the wrong number of arguments)? Could you whitelist only float-related operations in this branch?

Reply:
I guess it could, yes. 

I think if we want to avoid this special case (which really shouldn't be here), we should just forbid including arguments in the function name in general, then we wouldn't have to handle this case here. It's just that that was unproblematic when we were dealing only with strings, so I did it when making the factory for floating point operations. If we rewrite that code,  I think the only consequence is that users of floating point functions would always have to pass the rounding mode explicitly.